<hr>
<b><font color='red'>Ejercicio 01</font></b>  

Lee el archivo.txt 

In [1]:
f = open("archivo.txt", "r").readlines()

for string in f:
    a = string.split()
    for i in range(0,len(a),2):
        print("%.1f   %.2f" % (int(a[i]), int(a[i+1])))

1.0   2.00
2.0   4.00
3.0   6.00
4.0   8.00


y haz que tenga la siguiente salida:

    1.0   2.00  
    2.0   4.00  
    3.0   6.00  
    4.0   8.00  


<hr>
<b><font color='red'>Ejercicio 02</font></b>  

Haz lo mismo con el archivo archivo_error.txt, con cuidado, ya que a este archivo le falta algún dato: haz que cuando lo lea, si una línea no tiene dos columnas, no la escriba.

In [2]:
archivo="archivo_error.txt"
for line in open(archivo).readlines():
    a = line.split()
    if len(a) == 2:
        print("%.1f   %.2f" % (int(a[0]), int(a[1])))

1.0   2.00
3.0   6.00
4.0   8.00


<hr>
<b><font color='red'>Ejercicio 03</font></b>  

Lee el archivo.txt, haz que imprima las columnas en las filas y las filas en las columnas

In [3]:
archivo="archivo.txt"
text=open(archivo).readlines()
T=[]
for line in text:
    a=[]
    for i in line.split():
        a.append(i)
    T.append(a)
#print(T)
for i in range(len(T[0])):
    a=""
    for j in T:
        a=a+j[i]+" "
    print(a)

1 2 3 4 
2 4 6 8 


<hr>
<b><font color='red'>Ejercicio 04</font></b>  

1. Escribe en un archivo la siguiente matriz:
$
\begin{equation}
\begin{pmatrix}
3 & 1 & 3\\
-1 & 1 & 0\\
-2 & 4 & 1
\end{pmatrix}
\end{equation} 
$
2. Lee el archivo
3. Calcula con los datos leídos el determinante de la matriz y sácalo por pantalla

In [4]:
f = open("matriz.txt", "w")
f.write('3 1 3\n')
f.write('-1 1 0\n')
f.write('-2 4 1')
f.close()
f = open("matriz.txt", "r").readlines()
m = []
for line in f:
    aux=[]
    for i in line.split():
        aux.append(float(i))
    m.append(aux)
print("Resultado de la matriz: ", m)

i =(  m[0][0]*m[1][1]*m[2][2]
    + m[1][0]*m[2][1]*m[0][2]
    + m[2][0]*m[0][1]*m[1][2]
    - m[2][0]*m[1][1]*m[0][2]
    - m[0][0]*m[2][1]*m[1][2]
    - m[1][0]*m[0][1]*m[2][2])
    
print(i)

Resultado de la matriz:  [[3.0, 1.0, 3.0], [-1.0, 1.0, 0.0], [-2.0, 4.0, 1.0]]
-2.0


<hr>
<b><font color='red'>Ejercicio 05</font></b>  

En el criptoanálisis, la técnica de análisis de frecuencias consiste en el aprovechamiento de estudios sobre la frecuencia de las letras o grupos de letras en los idiomas para poder establecer hipótesis para aprovecharlas para poder descifrar un texto cifrado sin tener la clave de descifrado (romper). Es un método típico para romper cifrados clásicos. 
Haz un análisis de frecuencias para los 10 primeros capítulos del libro "el_quijote.txt"

In [5]:
abc="ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"

def frecuencias(archivo):
    f=[0]*len(abc)
    N=0
    for line in open(archivo).readlines():
        for word in line.split():
            for i in word.upper():
                a=i.replace("Á","A").replace("É","E").replace("Í","I").replace("Ó","O").replace("Ú","U")
                if a in abc:
                    f[abc.index(a)]=f[abc.index(a)]+1
                    N=N+1
    for k in range(len(f)):
        f[k]=f[k]/N
    return f

f=frecuencias("el_quijote.txt")

for i in abc:
    print("%.2c  " % i,"%.3f" % f[(abc.index(i))])

A   0.127
B   0.017
C   0.037
D   0.053
E   0.138
F   0.005
G   0.011
H   0.012
I   0.053
J   0.007
K   0.000
L   0.060
M   0.026
N   0.066
Ñ   0.000
O   0.099
P   0.021
Q   0.019
R   0.061
S   0.073
T   0.037
U   0.048
V   0.011
W   0.000
X   0.000
Y   0.015
Z   0.004


Salida:

    A   0.127
    B   0.017
    C   0.037
    D   0.053
    E   0.138
    F   0.005
    .
    .
    .

<hr>
<b><font color='red'>Ejercicio 06</font></b>  

Haz un análisis de frecuencias en el libro encriptado "RH RQCEPESEFM AR MPM.txt" y compáralo con el análisis anterior


In [6]:
f2=frecuencias("RH RQCEPESEFM AR MPM.txt")

print("el_quijote.txt","  RH RQCEPESEFM AR MPM.txt")    
for i in abc:
    print("%.2c  " % i,"%.3f" % f[(abc.index(i))],"       %.2c  " % i,"%.3f" % f2[(abc.index(i))])

el_quijote.txt   RH RQCEPESEFM AR MPM.txt
A   0.127        A   0.049
B   0.017        B   0.006
C   0.037        C   0.040
D   0.053        D   0.061
E   0.138        E   0.126
F   0.005        F   0.007
G   0.011        G   0.000
H   0.012        H   0.052
I   0.053        I   0.032
J   0.007        J   0.012
K   0.000        K   0.068
L   0.060        L   0.001
M   0.026        M   0.093
N   0.066        N   0.027
Ñ   0.000        Ñ   0.011
O   0.099        O   0.010
P   0.021        P   0.066
Q   0.019        Q   0.069
R   0.061        R   0.136
S   0.073        S   0.017
T   0.037        T   0.044
U   0.048        U   0.045
V   0.011        V   0.009
W   0.000        W   0.000
X   0.000        X   0.002
Y   0.015        Y   0.009
Z   0.004        Z   0.004


salida:

    el_quijote.txt   RH RQCEPESEFM AR MPM.txt
    A   0.127        A   0.049
    B   0.017        B   0.006
    C   0.037        C   0.040
    .                .
    .                .
    .                .

<hr>
<b><font color='red'>Ejercicio 07</font></b>  

Comparando el análisis de frecuencias, ¿sabrías decir cuál es el título del libro?  
RH RQCEPESEFM AR MPM

**Solución:** comparando las frecuencias se deduce la sustitución: la letra más frecuente del texto cifrado (R, 0.136) corresponde a la más frecuente en español (E, 0.138); la segunda (E, 0.126) a la A (0.127); M (0.093) a la O (0.099), etc. Aplicando la sustitución al título:

    RH RQCEPESEFM AR MPM  →  EL ESCARABAJO DE ORO

Se trata de *El escarabajo de oro*, de Edgar Allan Poe.

<hr>
<b><font color='red'>Generar el ejercicio "El escarabajo de oro"</font></b>

En este cuento de Edgar Allan Poe, el protagonista (el señor Legrand) encuentra un tesoro gracias a que es capaz de descifrar un texto en clave con las pistas suficientes para ir en su busca.  

El señor Legrand consigue descifrar el texto cifrado porque va sustituyendo cada signo (cifra o símbolo) por una letra del abecedario, teniendo en cuenta que los signos que más se repiten deben corresponder, más o menos, a las letras que más se usan en español: la E, la A, la O... Después del mensaje en clave puedes ver, en rojo, el abecedario ordenado por orden de frecuencia: la letra más usada es la E, seguida de la A, etc.

In [7]:
abc="ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"
rem="ESCARBJODFGHIKLMNÑPQTUVWXYZ"

texto=""
archivo="El_escarabajo_de_oro_SOL.txt"
for line in open(archivo).readlines():
    for word in line.split():
        for i in word.upper():
            a=i.replace("Á","A").replace("É","E").replace("Í","I").replace("Ó","O").replace("Ú","U")
            if a in abc:
                texto=texto+rem[abc.index(a)]
            else:
                texto=texto+i
        texto=texto+' '
    texto=texto+'\n'

with open("RH RQCEPESEFM AR MPM.txt", "w") as fout:
    fout.write(texto)